src/yaml/yamlUtils.ts

```diff
@@ -308,3 +311,148 @@ export function convertJsonSnippetToYaml(snippetText: string) {
         })
         .join('\n')
 }
+
+function getNumberOfLeftSpaces(text: string) {
+    return text.length - text.trimLeft().length
+}
+
+function getBackwardOffsetData(text: string, offset: number, initialNumberOfSpaces: number) {
+    let isDirectChildOfStates = false
+    let isGrandChildOfStates = false
+    let isWithinCatchRetryState = false;
+    let hasCatchPropSibling = false;
+    let hasRetryPropSibling = false;
+    let beginLineOffset = offset;
+    let levelsDown = 0;
+
+    for (let i = offset; i >= 0; i--) {
+        if (text[i] === '\n') {
+            const lineText = text.slice(i + 1, beginLineOffset)
+            const numberOfPrecedingSpaces = getNumberOfLeftSpaces(lineText)
+            const trimmedLine = lineText.trim()
+            beginLineOffset = i
+
+            // Ignore empty lines
+            if (trimmedLine.length === 0) {
+                continue
+            }
+
+            // If number of spaces lower than that of the cursor
+            // it is a parent property or a sibling of parent property
+            if (numberOfPrecedingSpaces < initialNumberOfSpaces) {
+                if (trimmedLine.startsWith('States:')) {
+                    isDirectChildOfStates = levelsDown === 0;
+                    isGrandChildOfStates = levelsDown === 1;
+                    break
+                } else if (levelsDown === 0) {
+                    levelsDown++
+                    continue
+                } else {
+                    break
+                }
+
+            // If number of spaces is higher than that of the cursor it means it is a child
+            // of the property or of its siblings
+            } else if (numberOfPrecedingSpaces > initialNumberOfSpaces) {
+                continue
+            } else if (levelsDown > 0) {
+                continue
+            }
+
+            hasCatchPropSibling = trimmedLine.startsWith('Catch:') || hasCatchPropSibling
+            hasRetryPropSibling = trimmedLine.startsWith('Retry:') || hasRetryPropSibling
+
+            const isCatchRetryState = CATCH_RETRY_STATE_REGEX.test(trimmedLine)
+
+            if (isCatchRetryState) {
+                isWithinCatchRetryState = true
```

Comment:
Can we directly assign the value from the regex to the variable with an OR operator?

Reply:
So we want to only update this variable if it is true but don't want make the true value back to false. It is a logical latch. I could use OR operator here to get same functionality but I don't think it brings any improvement to code clarity.